RedditSharp/WebAgent.cs

```diff
@@ -112,13 +66,38 @@ public int RequestsThisBurst
         public WebAgent( string accessToken ) {
             RootDomain = OAuthDomainUrl;
             AccessToken = accessToken;
+            RateLimitUsed = 0;
+            RateLimitReset = DateTimeOffset.UtcNow;
+            RateLimitRemaining = IsOAuth ? 60 : 30;
         }
 
         /// <inheritdoc />
-        public virtual async Task<JToken> ExecuteRequestAsync(HttpRequestMessage request)
+        public virtual async Task<JToken> ExecuteRequestAsync(Func<HttpRequestMessage> request)
         {
-            await EnforceRateLimit().ConfigureAwait(false);
-            var response = await _httpClient.SendAsync(request).ConfigureAwait(false);
+            if (request == null)
+              throw new ArgumentNullException(nameof(request));
+            const int maxTries = 20;
+            HttpResponseMessage response;
+            var tries = 0;
+            do {
+              await rateLimitLock.WaitAsync().ConfigureAwait(false);
```

Comment:
As you brought up in the other version, this does essentially single thread the web requests though right? Maybe a better option would be to have some sort of toggle to either be "safer" or add some sort of counter to try and pre-emptively assume any request will count towards the limit, but update afterwards when the header responds. Whichever header responds last "should" be accurate as that would be the last executed request?